src/lib/zexe_backend/zexe_backend_common/plonk_plookup_constraint_system.ml

```diff
@@ -0,0 +1,1172 @@
+    | None, [] ->
+        (Fp.zero, `Constant)
+    | _, t0 :: terms -> (
+        let terms =
+          let acc, i, ts, _ = accumulate_sorted_terms t0 terms in
+          List.rev ((acc, i) :: ts)
+        in
+        match terms with
+        | [] ->
+            assert false
+        | [(ls, lx)] -> (
+          match constant with
+          | None ->
+              (ls, `Var (V.External lx))
+          | Some c ->
+              (* res = ls * lx + c *)
+              let res = create_internal ~constant:c sys [(ls, External lx)] in
+              add_generic_constraint ~l:(External lx) ~o:res
+                [|ls; Fp.zero; Fp.(negate one); Fp.zero; Fp.zero; Fp.zero; c|]
+                (* Could be here *)
+                sys ;
+              (Fp.one, `Var res) )
+        | (ls, lx) :: tl ->
+            let rs, rx = completely_reduce sys tl in
+            let res =
+              create_internal ?constant sys [(ls, External lx); (rs, rx)]
+            in
+            (* res = ls * lx + rs * rx + c *)
+            add_generic_constraint ~l:(External lx) ~r:rx ~o:res
+              [| ls
+               ; rs
+               ; Fp.(negate one)
+               ; Fp.zero
+               ; Fp.zero
+               ; Fp.zero
+               ; (match constant with Some x -> x | None -> Fp.zero) |]
+              (* Could be here *)
+              sys ;
+            (Fp.one, `Var res) )
+
+  let add_constraint ?label:_ sys
+      (constr :
+        ( Fp.t Snarky_backendless.Cvar.t
+        , Fp.t )
+        Snarky_backendless.Constraint.basic) =
+    (*
+    let deb = Sexp.to_string ([%sexp_of: (Fp.t Snarky_backendless.Cvar.t, Fp.t) Snarky_backendless.Constraint.basic] constr) in
+    print_endline deb;
+*)
+    sys.hash <- feed_constraint sys.hash constr ;
+    let red = reduce_lincom sys in
+    let reduce_to_v (x : Fp.t Snarky_backendless.Cvar.t) : V.t =
+      let s, x = red x in
+      match x with
+      | `Var x ->
+          if Fp.equal s Fp.one then x
+          else
+            let sx = create_internal sys [(s, x)] in
+            (* s * x - sx = 0 *)
+            add_generic_constraint ~l:x ~o:sx
+              [| s
+               ; Fp.zero
+               ; Fp.(negate one)
+               ; Fp.zero
+               ; Fp.zero
+               ; Fp.zero
+               ; Fp.zero |]
+              sys ;
+            sx
+      | `Constant ->
+          let x = create_internal sys ~constant:s [] in
+          add_generic_constraint ~l:x
+            [|Fp.one; Fp.zero; Fp.zero; Fp.zero; Fp.zero; Fp.zero; Fp.negate s|]
+            sys ;
+          x
+    in
+    match constr with
+    | Snarky_backendless.Constraint.Square (v1, v2) -> (
+        let (sl, xl), (so, xo) = (red v1, red v2) in
+        match (xl, xo) with
+        | `Var xl, `Var xo ->
+            (* (sl * xl)^2 = so * xo
+               sl^2 * xl * xl - so * xo = 0
+            *)
+            add_generic_constraint ~l:xl ~r:xl ~o:xo
+              [| Fp.zero
```

Comment:
It would be nice (more readable, take up less space) if these were records, and these overrode fields on a default like `{generic_zero with o= Fp.negate so; m= Fp.(s1 * s1)}`. This should perhaps go in the 5-wires implementation if we're combining them, however.